alf/metrics/metrics.py

```diff
@@ -18,57 +18,44 @@
 """
 import torch
 
+import alf
+import alf.utils.data_buffer as db
+
 from . import metric
 
 
-class THDeque(torch.nn.Module):
-    """Torch (TH) deque backed by torch.tensor storage."""
+class MetricBuffer(torch.nn.Module):
+    """A metric buffer for computing average metric values. The buffer is assumed
+    to store only scalar values."""
 
     def __init__(self, max_len, dtype):
-        """Constructor.
-
+        """
         Args:
-            max_len (int): maximum length of the deque
-            dtype (torch.dtype): dtype of the content of the deque
+            max_len (int): maximum length of the buffer
+            dtype (torch.dtype): dtype of the content of the buffer
         """
         super().__init__()
-        shape = (max_len, )
-        self.dtype = dtype
-        self.register_buffer('_max_len',
-                             torch.tensor(max_len, dtype=torch.int64))
-        self.register_buffer('_buffer', torch.zeros(shape, dtype=dtype))
-        self.register_buffer('_head', torch.zeros((), dtype=torch.int64))
-
-    @property
-    def data(self):
-        return self._buffer[:self.length]
+        self._dtype = dtype
+        self._buf = db.DataBuffer(
+            data_spec=alf.tensor_specs.TensorSpec((), dtype=dtype),
+            capacity=max_len,
+            device=alf.get_default_device())
 
     def append(self, value):
-        """Appends value to deque.
-        Overwrites earliest value if deque exceeds max_len.
+        """Append multiple values to the buffer.
 
         Args:
-            value (scalar or tensor of dtype): input value to collect
-
-        Returns:
-            Nothing
+            value (Tensor): a batch of scalars with the shape :math:`[B]`.
         """
-        position = torch.remainder(self._head, self._max_len)
-        self._buffer[position] = value
-        self._head.add_(1)
+        self._buf.add_batch(value)
 
-    @property
-    def length(self):
-        return torch.min(self._head, self._max_len)
+    def mean(self):
+        if self._buf.current_size == 0:  # avoid nan
+            return torch.as_tensor(0, dtype=self._dtype)
+        return self._buf.get_all()[:self._buf.current_size].mean()
```

Comment:
`self._buf.get_batch(self._buf.current_size).mean()` is perhaps safer, in case some entries get deleted from buffer.

get_all simply returns the whole buffer, and probably shouldn't be exposed for external use.

Reply:
get_batch will randomly sample with replacement. Here I need to average over all elements in the buffer. And DataBuffer can only overflow data but not delete data.

Reply:
You are right, let's leave as is then.